public_html/wp-content/plugins/pattern-creator/src/plugins/viewport-header-control.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * External Dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import {
+	DropdownMenu,
+	MenuGroup,
+	MenuItem,
+	TextControl,
+} from '@wordpress/components';
+import { registerPlugin } from '@wordpress/plugins';
+
+/**
+ * Internal dependencies
+ */
+const CONTAINER_ID = 'pattern-viewport-container';
+
+const WIDTH_DEFAULTS = [
+	{
+		label: 'Narrow (320px)',
+		value: '320px',
+	},
+	{
+		label: 'Default (800px)',
+		value: '800px',
+	},
+	{
+		label: 'Full (100%)',
+		value: '100%',
+	},
+];
+
+/**
+ * Removes the button container if it exists
+ */
+const maybeRemoveContainer = () => {
+	const container = document.getElementById( CONTAINER_ID );
+
+	if ( container && container.parentElement ) {
+		container.parentElement.removeChild( container );
+	}
+};
+
+const insertContainer = ( btnDomRef ) => {
+	const container = document.createElement( 'div' );
+	container.id = CONTAINER_ID;
+
+	btnDomRef.parentElement.insertBefore( container, btnDomRef );
+};
+
+/**
+ * Insert an html element to the right.
+ *
+ * @param {HTMLElement} newNode Element to be added
+ */
+export const insertButton = ( newNode ) => {
+	// The Gutenberg Publish Button
+	const btnDomRef = document.querySelector( '.editor-post-publish-button__button' );
+
+	if ( ! btnDomRef ) {
+		return;
+	}
+
+	// We may re-insert the same button if it state's changes
+	maybeRemoveContainer();
+
+	insertContainer( btnDomRef );
+
+	/* eslint-disable no-undef*/
+	ReactDOM.render( newNode, document.getElementById( CONTAINER_ID ) );
+};
+
+const ViewportHeaderControl = () => {
+	const { editPost } = useDispatch( 'core/editor' );
+	const postMetaData = useSelect( ( select ) => select( 'core/editor' ).getEditedPostAttribute( 'meta' ) || {} );
+	const [ viewportWidth, setViewport ] = useState( postMetaData.wpop_viewport_width );
+
+	const updateElementWidth = () => {
+		const page = document.querySelector( '.block-editor-block-list__layout' );
+		page.style.width = viewportWidth;
+	};
+
+	useEffect( () => {
+		editPost( {
+			meta: {
+				...postMetaData,
+				wpop_viewport_width: viewportWidth,
+			},
+		} );
+
+		// Update the main container width
+		updateElementWidth();
+
+		insertButton(
+			<DropdownMenu
+				className="viewport-header-control"
+				icon={ null }
+				text={ sprintf(
+					/* translators: %s viewport width as css, ie: 100% */
+					__( `Width (%s)`, 'wporg-patterns' ),
+					viewportWidth
+				) }
+				label={ __( 'Select a viewport width', 'wporg-patterns' ) }
+			>
+				{ ( { onClose } ) => (
+					<>
+						<MenuGroup label="Viewport Width">
+							<p className="viewport-header-control__copy">
+								This is used when displaying a preview of your pattern.
+							</p>
+							<TextControl
```

Comment:
I wanted to use the `UnitControl` component here but since we are injecting the button into the dom via the `insertButton` function, we can't manage the state properly. Every time the state updates, in this case, `viewportWidth`, the button is removed and rebound to the DOM. So technically we can only watch for changes in one variable. The way the `<UnitControl>` works, it requires you to manage both the number and the unit separately.

The alternative to injecting the button is to `pin` an icon into the header and rearrange the order using CSS to match the order in the designs.  I looked into that and it appears you can't just pin an icon. You have to provide a `PluginSidebar` which we don't want. We only want a dropdown menu.